### ISSUE TITLE
Duplicate ID false flag with Vue component (attachToDocument)

### ISSUE THREAD

**Issue by CrucifixArnaud** [nickcolley/jest-axe#56]
Hello, I'm not sure it's directly an issue on our side, but I will ask. 

I'm using jest-axe to test Vue component (with vue-test-utils), and since the update of jest-axe v3.1.1, it return errors for duplicate id ("id attribute value must be unique (duplicate-id)") for some of my component. But after verification, the ID are not duplicated, it seem that these false errors are only throw when mounting the component with the option attachToDocument set to true.

Here are a simple example: 

testId.vue
```
<template>
    <div>
        <p id="test-id">Test Id</p>
    </div>
</template>
```
testId.spec.js
```
import Vue from "vue";
import { shallowMount} from "@vue/test-utils";

import testId from "testId.vue";

// Use axe accessibility tools
// https://www.deque.com/axe/
// https://github.com/nickcolley/jest-axe
import { axe, toHaveNoViolations } from "jest-axe";
expect.extend(toHaveNoViolations);

describe("CardList.vue", () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallowMount(testId, {
            attachToDocument: true,
        });
    });

    describe('Mounted', () => {
        test("It should not contain any axe accessibility violation", async () => {
            const html = wrapper.html();

            expect(await axe(html)).toHaveNoViolations()
        });
    });
});
```

Return: 

```
Expected the HTML found at $('div:nth-child(2) > div
                                                     
<p id="test-id">Test Id</p>                         
                                                     
Received:                                           
                                                    
"id attribute value must be unique (duplicate-id)"  
```

Any idea ? Thank you

**Comment by nerfologist** [comment]
Hello, I think we are suffering from the same or a similar problem with `jest-axe` 3.1.1. **It does not occur on jest-axe 3.1.0**, to which we are sticking for now.

Our simple input component, whose `debug()` output (from `react-testing-library`) is as follows:
```html
    <body>
      <div>
        <div
          class="sc-iwsKbI gFRpPd"
        >
          <label
            class="sc-bdVaJa hIvQaX"
            for="mocked-shortid-7"
          >
            A label
            <span
              class="sc-bxivhb dUKaDy"
            >
              (
              optional
              )
            </span>
          </label>
          <div
            class="sc-EHOje eQMpHY"
          >
            <input
              class="sc-ifAKCX hLVQBy"
              id="mocked-shortid-7"
              type="text"
            />
          </div>
        </div>
      </div>
    </body>
```
...and which we're testing this way:
```javascript
import { axe } from 'jest-axe';

describe('accessibility', () => {
  it('has no a11y violations', async () => {
    const { container } = render(getComponent(props));
    expect(await axe(container.innerHTML)).toHaveNoViolations();
  });
});
```

...reports the following errors:
```
    Expected the HTML found at $('div:nth-child(2) > .sc-iwsKbI.gFRpPd > .sc-EHOje.eQMpHY > input') to have no violations:
    <input class="sc-ifAKCX hLVQBy" id="mocked-shortid-7" type="text">
    Received:
    "IDs used in ARIA and labels must be unique (duplicate-id-aria)"

    Expected the HTML found at $('div:nth-child(2) > .sc-iwsKbI.gFRpPd > .sc-EHOje.eQMpHY > input') to have no violations:
    <input class="sc-ifAKCX hLVQBy" id="mocked-shortid-7" type="text">
    Received:
    "Form elements must have labels (label)"
```

The debug output tests OK in the [axe playground](https://www.deque.com/axe/playground/).

**Comment by nickcolley** [comment]
Thanks for the test case! Sorry for slowness from me, I'm a bit overwhelmed right now

**Comment by nerfologist** [comment]
Hey, I added a CodeSandbox with a failing test: https://codesandbox.io/s/jpjozryn7v.

**Comment by nickcolley** [comment]
Thanks, we can add that to the test suite and go from there... Will try and do this soon.

**Comment by donifer** [comment]
Having the same issue as well

**Comment by deini** [comment]
Been poking around this issue today and I believe I know what's going on. Both `codesandbox` examples (here and in https://github.com/nickcolley/jest-axe/issues/57) are using `react-testing-library`. I'm assuming `@vue/test-utils` does something similar.

When doing something like:
```js
const { container } = render(<MyComponent />);
const results = await axe(container.innerHTML);

expect(results).toHaveNoViolations();
```

`react-testing-library` creates a container and adds it to `document.body`, you can see it [here](https://github.com/kentcdodds/react-testing-library/blob/a33ccf8f52ecffd46a7ee2438454949794270f67/src/index.js#L26).

Then `jest-axe` creates another container and also appends it to `document.body` [here](https://github.com/nickcolley/jest-axe/blob/37384ae90c93723c85785bc0bdb96a4b3e512766/index.js#L33:L37).

So we end up with two rendered versions of `<MyComponent />` in `document.body` which is why `axe-core` is throwing the duplicate ids error.

### Workaround <span>#</span>1
```js
const div = document.createElement('div');
const { container } = render(<MyComponent />, { container: div });
const results = await axe(container.innerHTML);

expect(results).toHaveNoViolations();
```

Pass an element to the `container` option of react-testing-library's `render()` so that it doesn't add it to the `document` at all.

### Workaround <span>#</span>2
```js
import { cleanup, render } from 'react-testing-library';

const { container } = render(<MyComponent />, { container: div });
const html = container.innerHTML;

cleanup();

const results = await axe(container.innerHTML);

expect(results).toHaveNoViolations();
```

Store the html string and call `cleanup()` so that `react-testing-library` unmounts its container. (I don't like this workaround tho).

### Possible solutions:
- Leave it as-is and add some documentation around it.
- Save document.body nodes like @thomasheyenbrock was doing initially in his PR and restore them. But might introduce [unexpected behavior](https://github.com/nickcolley/jest-axe/pull/52#discussion_r253835340).
- Something else I haven't thought about (very likely!).

Would love to hear your thoughts on this one @thomasheyenbrock @nickcolley

**Comment by thomasheyenbrock** [comment]
Well, my changes definitively are the reason for this issue, sorry folks! 🙈 

For me it doesn't feel good after all to modify the existing `document`. One way or another this leads to issues as we can see here.

A question out of curiosity (probably for @nickcolley): Why do we need to pass a string to `axe`? Why not pass a DOM node that should be checked for a11y? After all `axe-core` also needs a DOM element. 

Then we could just use it like so:
```js
const { container } = render(<span>some text</span>)
const results = await axe(container)
expect(results).toHaveNoViolations()
```

Then there wouldn't be any need to modify the `document` anymore which would resolve this issue and also #50.

**Comment by dakebl** [comment]
Hello all, I've add PR #62 to try and add the functionality mentioned by @thomasheyenbrock mentioned.

I've also added the ability to pass in react elements, with axe converting valid react elements to string content under the hood:

```javascript
// react-testing-library example
const { container } = render(React.createElement('img', { src: '#' }))
const results = await axe(container)      
expect(results).toHaveNoViolations()

// react element example
const element = React.createElement('img', { src: '#' })
const results = await axe(element)      
expect(results).toHaveNoViolations()
```

What do you think @nickcolley?

**Comment by citypaul** [comment]
This is still an issue at the time of writing. It's quite a big one for users of react testing library, as it appears  to also have an impact on existing tests, causing good tests to fail under some circumstances.

Here's an example:

```jsx
  test("Will render a badge with specified text and id", () => {
    const badgeText = "PH"
    const id = "some_id"
    const { getByText } = render(<Badge id={id} text={badgeText} />)

    expect(getByText(badgeText)).toHaveAttribute("id", id)
  })
```

The above test renders this DOM output if I don't pull in `jest-axe`:

```html
<body>
        <div>
          <span
            class="c-badge"
            id="some_id"
          >
            PH
          </span>
        </div>
</body>
```

However, as noted above, `jest-axe` is modifying the DOM, which results in the component appearing twice in the DOM in my tests now (and this isn't just tests where I'm asserting the `jest-axe` output - this is any test that happens to pull in the `jest-axe` library):

```html
<body>
      <span
        class="c-badge"
        id="some_id22"
      >
        PH
      </span>
      <div>
        <span
          class="c-badge"
          id="some_id"
        >
          PH
        </span>
      </div>
</body>
```

This causes my test to now fail: 

```
  ● Atoms > Badge › Will render a badge with specified text and id

    Found multiple elements with the text: PH

    (If this is intentional, then use the `*AllBy*` variant of the query (like `queryAllByText`, `getAllByText`, or `findAllByText`)).
```

The solution noted above looks good, and this PR looks like it'll do the trick, but currently the build is failing there: https://github.com/nickcolley/jest-axe/pull/62

**Comment by dakebl** [comment]
I fixed the build issue now. Thanks for the spot @citypaul!
What do you think @nickcolley?

**Comment by kentcdodds** [comment]
I've opened a PR #70 that should address this really well. Just pass a DOM node to jest-axe, and it'll work perfectly :)

Feedback and 👍 on the PR are appreciated.

**Comment by nickcolley** [comment]
Published the contributed fix as 3.2.0, going to close this out as I believe it fixes this issue, please raise another issue if you run into any more issues.

Thanks to @dakebl for contributing and everyone else helping me with the reviewing!
